### PR TITLE
fix: fix layout matching

### PR DESCRIPTION
**Pull request by caalador**
The automatic layout matching
is mistakenly made as equals
when it should be starts with
path parts.

Fixes vaadin/flow#20072

### CODE REVIEW

File: packages/ts/file-router/src/vite-plugin/applyLayouts.ts [hunk 1]
```diff
@@ -26,6 +26,36 @@ function enableFlowLayout(route: RouteMeta): RouteMeta {
...
+      const splitLayout: string[] = layout.split('/');
+      if (splitLayout.length > splitPath.length) {
+        return false;
+      }
+      for (let i = 0; i < splitLayout.length; i++) {
+        if (!(splitPath[i] === splitLayout[i])) {
+          return false;
+        }
+      }
+      return true;
```
Comment (taefi) [inline packages/ts/file-router/src/vite-plugin/applyLayouts.ts:54]: Isn't this the same as `path.startWith(layout)`?
Another thing, can layouts have patterns e.g. `/home/*`? If so, I believe `/home/*` should match a path like `/home/contacts` but the current logic will return false, if I understand correctly. The trailing `/` and `possible `/*` should also be stripped before this logic is checked.

Comment (caalador) [reply packages/ts/file-router/src/vite-plugin/applyLayouts.ts:54]: The path parts are split with / so there should be not / in any part. Except if javascript split works in mysterious ways.

No patterns as you only give the path it matches to so all given layout paths are seen as ending with /*

Comment (caalador) [reply packages/ts/file-router/src/vite-plugin/applyLayouts.ts:54]: Also it is not the same as startWith as if the layout path is `/home/owner` it should not match `/home/owner2/house`, but should match `/home/owner/house`

File: packages/ts/file-router/src/vite-plugin/applyLayouts.ts [hunk 2]
```diff
@@ -26,6 +26,36 @@ function enableFlowLayout(route: RouteMeta): RouteMeta {
...
+ * and '/home' matches '/home' anything with the start path '/home/*'
+ *
+ * @param layoutPaths - available layout paths
+ * @param path - to check for layout
+ */
+function layoutExists(layoutPaths: string[], path: string) {
+  const splitPath: string[] = path.split('/');
+
+  return (
+    layoutPaths.filter((layout) => {
```
[the remark this answers is not in the document]
Comment (caalador) [reply packages/ts/file-router/src/vite-plugin/applyLayouts.ts]: done

File: packages/ts/file-router/src/vite-plugin/applyLayouts.ts [hunk 3]
```diff
@@ -26,6 +26,34 @@ function enableFlowLayout(route: RouteMeta): RouteMeta {
...
+  return layoutPaths.some((layout) => {
+    if (layout.length === 0) {
+      return true;
+    }
+    const splitLayout: string[] = layout.split('/');
+    if (splitLayout.length > splitPath.length) {
+      return false;
+    }
+    for (let i = 0; i < splitLayout.length; i++) {
+      if (!(splitPath[i] === splitLayout[i])) {
```
Comment (taefi) [inline packages/ts/file-router/src/vite-plugin/applyLayouts.ts:49]: nit:
```suggestion
      if (splitPath[i] !== splitLayout[i]) {
```

**Review by taefi — APPROVED** [review]

**Review by taefi — APPROVED** [review]
LGTM

**Pull request merged**